### PR TITLE
SOCI cleanup.

### PR DESCRIPTION
**Pull request by rec**
Reviewers: 

@HowardHinnant @seelabs

### CODE REVIEW

**Comment by HowardHinnant** [comment]
Nice job.  Most of my comments are minor and can be ignored.  I think the most important one is getting a definitive answer on the question you asked in your comments regarding `if ((1 + pages) % checkpointPageCount)`.

File: src/ripple/app/data/SociDB.cpp [hunk 1]
```diff
@@ -51,24 +54,21 @@ getSociInit (BasicConfig const& config,
...
-    {
-        std::string const path = config.legacy ("database_path");
-        std::string const ext =
-            (dbName == "validators" || dbName == "peerfinder") ? ".sqlite"
-                                                               : ".db";
-        return detail::getSociSqliteInit(dbName, path, ext);
-    }
-    else
-    {
+    if (backendName != "sqlite")
```
Comment (seelabs) [inline src/ripple/app/data/SociDB.cpp]: New backends will be added as `else if (backendName == "postgresql") ...`. For that reason I prefer the original version of this code, even if your version is nicer when we only have one backend.

File: src/ripple/app/data/DatabaseCon.h [hunk 2]
```diff
@@ -85,27 +85,27 @@ class DatabaseCon
...
 private:
-    std::unique_ptr<WALCheckpointer> walCheckpointer_;
-    // shared_ptr to handle lifetime issues with walCheckpointer_
-    // never null.
-    std::shared_ptr<soci::session> session_;
     LockedSociSession::mutex lock_;
+
+    soci::session session_;
+    std::unique_ptr<Checkpointer> checkpointer_;
 };
```
Comment (seelabs) [inline src/ripple/app/data/DatabaseCon.h]: If we swap the declaration order of `session_` and `checkpointer_` then `checkpointer_` will hold a stale ref to `session_` when `DatabaseCon` is destroyed. I'm OK getting rid of the `shared_ptr` that prevented this, but we should add a comment that declaration order matters here.

File: src/ripple/app/data/SociDB.cpp [hunk 3]
```diff
@@ -27,12 +27,15 @@
 #include <boost/filesystem.hpp>
 
 namespace ripple {
+
+static auto checkpointPageCount = 1000;
```
Comment (seelabs) [inline src/ripple/app/data/SociDB.cpp]: Put `checkpointPageCount` into the detail namespace.

File: src/ripple/app/data/SociDB.cpp [hunk 4]
```diff
@@ -84,55 +84,63 @@ std::string SociConfig::connectionString () const
...
-    }
     else
-    {
         throw std::runtime_error ("Unsupported soci backend: " + beName);
-    }
+}
+
+static
+sqlite_api::sqlite3* getConnection (soci::session& s)
+{
```
Comment (seelabs) [inline src/ripple/app/data/SociDB.cpp]: Put into `detail` namespace.

File: src/ripple/app/data/SociDB.cpp [hunk 5]
```diff
@@ -140,84 +148,99 @@ void convert(soci::blob& from, std::vector<std::uint8_t>& to)
...
     if (!from.empty ())
         to.write (0, reinterpret_cast<char const*>(&from[0]), from.size ());
 }
 
-int SqliteWALHook (void* s,
-                   sqlite_api::sqlite3*,
-                   const char* dbName,
-                   int walSize)
+namespace {
+
```
Comment (seelabs) [inline src/ripple/app/data/SociDB.cpp]: I thought anon namespaces were not allowed in rippled (I've had reviews comment when I've used them in other code). I think practice is to use the `ripple::detail` namespace for things like `WALCheckpointer`.

File: src/ripple/app/data/SociDB.cpp [hunk 6]
```diff
@@ -140,84 +148,99 @@ void convert(soci::blob& from, std::vector<std::uint8_t>& to)
...
-    sqlite_api::sqlite3_wal_hook (conn_, SqliteWALHook, this);
-}
-
-void WALCheckpointer::doHook (const char* db, int pages)
+void WALCheckpointer::runCheckpoint (const char* db, int pages)
 {
-    if (!conn_) return;
-
-    if (pages < 1000)
+    if ((pages + 1) % checkpointPageCount)
```
Comment (seelabs) [inline src/ripple/app/data/SociDB.cpp]: I don't believe this is correct. This needs to change back to the original `if (pages < checkpointPageCount)`. I left a comment in a previous commit explaining why (pages is the size of the wal file, which is consumed in a checkpoint).

File: src/ripple/app/data/SociDB.cpp (hunk 3 above)
Comment (vinniefalco) [inline src/ripple/app/data/SociDB.cpp]: Not necessary in a .cpp

File: src/ripple/app/data/SociDB.cpp (hunk 4 above)
Comment (vinniefalco) [inline src/ripple/app/data/SociDB.cpp]: Also not necessary in a .cpp

File: src/ripple/app/data/SociDB.cpp (hunk 5 above)
Comment (vinniefalco) [inline src/ripple/app/data/SociDB.cpp]: Its not so much that they are not allowed as that they have less effectiveness in a unity build. If WALCheckpointer is declared in a .cpp file, that's good enough. No need to add additional adornments.

File: src/ripple/app/data/SociDB.cpp (hunk 4 above)
Comment (seelabs) [inline src/ripple/app/data/SociDB.cpp]: @vinniefalco Thanks for the guidance. I'm making a note that new elements in cpp files should prefer to use the current top-level namespace (`ripple`) in cpp files and not use `detail` namespaces or anon namespaces.

**Comment by HowardHinnant** [comment]
:+1:

**Comment by rec** [comment]
Thanks!

I'm going to leave the `static` statements in - I think they provide documentary value that the names aren't supposed to be exposed to the rest of the world, even though they don't provide protection from collisions.

**Comment by vinniefalco** [comment]
@rec I agree (RE: `static`)

**Comment by seelabs** [comment]
:+1:

**Comment by nbougalis** [comment]
Merged as 03d1c0e and e874a26

**Pull request closed without merging**